### PR TITLE
grpc: reorganize grpc

### PR DESCRIPTION
**Pull request by JonasVautherin**
* The protos are now in a separate repo (DroneCore-Proto)
* DroneCore-Proto is a submodule in grpc/server
* gRPC is built automatically from cmake
* python_client disappears from this repo and goes in DroneCore-Python

I tested and could compile/run the server on Ubuntu, Arch Linux and Alpine Linux.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -11,3 +11,6 @@
 [submodule "libs/tinyxml2"]
 	path = libs/tinyxml2
 	url = https://github.com/leethomason/tinyxml2.git
+[submodule "grpc/server/proto"]
```
Comment (shakthi-prashanth-m) [inline .gitmodules]: @JonasVautherin So we have now created DroneCore-Proto repo for maintaining `.proto` files. I couldn't figure out why it is done so. We can as well have this in `grpc/server/proto/` isn't it ?

Comment (shakthi-prashanth-m) [reply .gitmodules]: @JonasVautherin 
What I understand is that, we now have
- **DroneCore-Python**:  Repo that has Python-Client samples for DroneCore like `fly_mission.py`, `takeoff_land.py`, etc.

- **DroneCore-Proto**: Repo that contains `.proto` files alone to be used by all the language bindings.

We're going to have in future:
- **DroneCore-Android:** Repo that is going to have Android app samples for DroneCore like `FlyMission` app, etc.
Please feel free to correct me.

Comment (JonasVautherin) [reply .gitmodules]: You're perfectly right! The idea is to split the clients from the backend. And since we need to share the proto files between them, I figured having a repo used as a submodule would be nice.

I am working on DroneCore-Python and DroneCore-Java that I will push very soon to get your feedback :-).

Comment (shakthi-prashanth-m) [reply .gitmodules]: Okay makes sense to me.

File: grpc/server/cmake/helpers/build_target.cmake [hunk 2]
```diff
@@ -0,0 +1,15 @@
+function(build_target SRC_DIR TARGET_DIR)
+    file(MAKE_DIRECTORY ${TARGET_DIR}/build)
+
+    execute_process(
+        COMMAND ${CMAKE_COMMAND}
```
Comment (julianoes) [inline grpc/server/cmake/helpers/build_target.cmake]: I see. So you run cmake out of cmake.

File: .gitmodules [hunk 3]
```diff
@@ -11,3 +11,6 @@
...
+	path = grpc/server/proto
+	url = git@github.com:dronecore/DroneCore-Proto.git
```
Comment (julianoes) [inline .gitmodules]: Did you choose the ssh link here on purpose? Do you know if this works for anyone without github ssh key?

File: grpc/server/cmake/external/grpc/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,23 @@
+cmake_minimum_required(VERSION 2.8)
+
+project(external-grpc)
+
+include(ExternalProject)
```
Comment (julianoes) [inline grpc/server/cmake/external/grpc/CMakeLists.txt]: Where do external projects end up? Are they rmoeved when you do `make clean`? Do we need something like `make distclean` to remove everything, so we can have `make clean` to only remove dronecore itself?

File: grpc/server/src/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    dronecore_server
+    PUBLIC
+    ${CMAKE_BINARY_DIR}/grpc/server/src
+    ${PLUGINS_DIR}
+)
+
+target_link_libraries(
+    dronecore_server
+    dronecore
+    gRPC::grpc++
```
Comment (julianoes) [inline grpc/server/src/CMakeLists.txt]: I didn't know libraries could have `::`.

File: CMakeLists.txt [hunk 6]
```diff
@@ -302,29 +301,7 @@ if(NOT IOS AND NOT ANDROID)
...
-            message(STATUS "PROTOBUF_DIR not set, checking automatically for Protobuf")
-            find_package(Protobuf REQUIRED)
-        endif()
-
-        add_subdirectory(grpc/server)
-        add_subdirectory(grpc/python_client)
-    else()
-        message(STATUS "gRPC not found, skipping gRPC build")
-    endif()
+    add_subdirectory(grpc/server)
```
Comment (julianoes) [inline CMakeLists.txt]: Should we add a cmake option, so you need to enable or can disable the gRPC build?

File: grpc/server/cmake/external/cares/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+cmake_minimum_required(VERSION 2.8)
+
+project(external-cares)
+
+include(ExternalProject)
+
+ExternalProject_add(
+    cares
+    GIT_REPOSITORY https://github.com/c-ares/c-ares.git
```
Comment (julianoes) [inline grpc/server/cmake/external/cares/CMakeLists.txt]: Why is this needed? Doesn't gRPC already come with c-ares?

File: .gitmodules (hunk 3 above)
Comment (JonasVautherin) [reply .gitmodules]: Good point :-). Gonna change this.

File: grpc/server/cmake/helpers/build_target.cmake (hunk 2 above)
Comment (JonasVautherin) [reply grpc/server/cmake/helpers/build_target.cmake]: Yes

File: grpc/server/src/CMakeLists.txt (hunk 5 above)
Comment (JonasVautherin) [reply grpc/server/src/CMakeLists.txt]: I learnt that as well :-). That is coming from the targets defined in gRPCConfig.cmake.

File: grpc/server/cmake/external/grpc/CMakeLists.txt (hunk 4 above)
Comment (JonasVautherin) [reply grpc/server/cmake/external/grpc/CMakeLists.txt]: The target is defined a few lines below:

    -DCMAKE_INSTALL_PREFIX=$ENV{DRONECORESERVER_CARES_BINARY_DIR}

So it goes in `build/default/external` in the end. I am not sure what could be a good way to achieve what you suggest, though.

File: grpc/server/cmake/external/cares/CMakeLists.txt (hunk 7 above)
Comment (JonasVautherin) [reply grpc/server/cmake/external/cares/CMakeLists.txt]: Well, I had to add the options in `grpc/server/cmake/external/grpc/CMakeLists.txt` one by one because it was requesting them for some reason. But honestly I am not completely sure why it could not use the one from the grpc repo =/.

Actually, in the previous version (the current `develop` branch), it was required to install protobuf separately (and to set PROTOBUF_DIR). Not sure why exactly.

File: CMakeLists.txt (hunk 6 above)
Comment (JonasVautherin) [reply CMakeLists.txt]: Checking this.

File: grpc/server/cmake/external/cares/CMakeLists.txt (hunk 7 above)
Comment (JonasVautherin) [reply grpc/server/cmake/external/cares/CMakeLists.txt]: Well, I am actually checking this, and I think I needed the "PACKAGE" lines in order to "install" grpc. But I cannot reproduce it correctly.

File: CMakeLists.txt (hunk 6 above)
Comment (avinash-palleti) [reply CMakeLists.txt]: I agree with @julianoes on this.

File: grpc/server/CMakeLists.txt [hunk 8]
```diff
@@ -1,74 +1,11 @@
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 2.8)
```
Comment (avinash-palleti) [inline grpc/server/CMakeLists.txt]: I remember we faced issue while testing with older cmake version(not sure how old though), that's why it was kept specific to 2.8.12
Hope you have tested with 2.8.0 so that anything above 2.8 will work.

File: grpc/server/src/dronecore_server.cpp [hunk 9]
```diff
@@ -50,13 +50,15 @@ void RunServer()
...
     std::fstream file;
-    file.open("grpc/proto/plugins.conf");
+    file.open("grpc/server/src/plugins/plugins.conf");
     if (!file) {
         LogErr() << "Error in reading conf file";
         return;
     }
     std::vector<::grpc::Service *> list;
-${PLUGIN_MAP_APPEND_STRING}
+    map["action"] = &createInstances<ActionRPCImpl>;
```
Comment (avinash-palleti) [inline grpc/server/src/dronecore_server.cpp]: Any specific reason to again having this hardcoded plugin names ?

File: grpc/server/src/plugins/plugins.conf [hunk 10]
```diff
@@ -1,3 +1,3 @@
-telemetry
 action
```
Comment (avinash-palleti) [inline grpc/server/src/plugins/plugins.conf]: Is this conf file used somewhere in current implementation ?

File: grpc/server/CMakeLists.txt (hunk 8 above)
Comment (JonasVautherin) [reply grpc/server/CMakeLists.txt]: I can put it back to 2.8.12.

File: grpc/server/src/dronecore_server.cpp (hunk 9 above)
Comment (JonasVautherin) [reply grpc/server/src/dronecore_server.cpp]: Right now, the includes are hardcoded anyway, so having both is not consistent to me. I kept the simpler one. But the auto-generation is still in the git history, so we could easily go back to it.

My feeling is that in the end, if we really want plugins, we cannot keep them in-process in dronecore_server. More specifically, if we require the customers to build their own version of dronecore_server in order to use their set of plugins, then it starts to look like a fork (i.e. they will provide their own SDK to their customers). And in this case their are not really "plugins".

So if we want real plugins, they should run in their own process and have a way to register to dronecore_server, other than being detected by cmake and compiled this way. That's probably something we should discuss during the next call :-).

Comment (Rjasuja) [reply grpc/server/src/dronecore_server.cpp]: >Right now, the includes are hardcoded anyway, so having both is not consistent to me.
For includes also, we can have one more run-time variable like PLUGIN_MAP_APPEND_STRING and assign it in autogenerate_plugin_container.cmake.

Comment (JonasVautherin) [reply grpc/server/src/dronecore_server.cpp]: Yes, definitely.

Still, I'd like to discuss that, as I am not completely sure this is what we want to do in the end. To be honest, I am not a big fan of auto-generating source files, as it adds some magic to the build system.

Again, I believe that either we go towards real plugins, in which case the plugins register themselves to the server, or we go towards some kind of drivers, in which case customers have to compile their own SDK. In case we go for the latter, I'd rather have them understand which library they link and how than having this auto-generated.

I would say we keep this discussion for the next WG call, if you don't mind :-).

File: CMakeLists.txt (hunk 6 above)
Comment (JonasVautherin) [reply CMakeLists.txt]: I added a BUILD_DRONECORESERVER option (see last commit).

File: grpc/server/cmake/external/grpc/CMakeLists.txt (hunk 4 above)
Comment (JonasVautherin) [reply grpc/server/cmake/external/grpc/CMakeLists.txt]: I tried to add `make distclean` and modify `make clean` to keep `build/default/external` (I kept this change in a commit that can be squashed later).

I am not sure where the `default` in `build/default` comes from, though. Moreover, I am not sure if my command is understandable and if it works on all platforms:

    clean:
        @find build/default/ -mindepth 1 -path build/default/external -prune -o -exec rm -rf {} +

File: grpc/server/cmake/external/cares/CMakeLists.txt (hunk 7 above)
Comment (JonasVautherin) [reply grpc/server/cmake/external/cares/CMakeLists.txt]: So, I figured c-ares has to be compiled as a package because of this message:

> gRPC_INSTALL will be forced to FALSE because gRPC_CARES_PROVIDER is
  "module"

Moreover, if I don't put protobuf as a package, somehow `gRPCTargets.cmake` is not generated.

I want grpc to get installed in `build/default/external/grpc`, so I need them as packages. Maybe there is another way using the in-tree versions and without installing grpc, but in this case I am not sure if the targets would be generated...

The way I see it is that either way, c-ares and grpc are dependencies. Here we have control on the protobuf version we use, and this setup works. If you feel like it is much cleaner to use the in-tree dependencies of grpc, I can try it. But I am not sure it is worth it (especially because the way I do it in this PR is another way, and not a hack).

File: grpc/server/cmake/external/grpc/CMakeLists.txt (hunk 4 above)
Comment (julianoes) [reply grpc/server/cmake/external/grpc/CMakeLists.txt]: `default` was just for the "native" build, so not ios and not android.

File: grpc/server/src/dronecore_server.cpp (hunk 9 above)
Comment (julianoes) [reply grpc/server/src/dronecore_server.cpp]: I vote against compile time code generation as well. It has created quite a bit of confusion in the past and makes it hard to get into this project.

File: grpc/server/cmake/external/grpc/CMakeLists.txt (hunk 4 above)
Comment (JonasVautherin) [reply grpc/server/cmake/external/grpc/CMakeLists.txt]: Ok so maybe it is fine for now, and we'll have to see how we adapt that when we compile `dronecore_server` for Android and iOS

File: grpc/server/src/dronecore_server.cpp (hunk 9 above)
Comment (JonasVautherin) [reply grpc/server/src/dronecore_server.cpp]: I could not agree more! To me, code generation is the same as `#ifdef`s. It is **always** worth spending time doing everything we can for not doing it :D.
